### DISCUSSION TITLE
about FIPS-140 enabling

### DISCUSSION THREAD

**Discussion by jyao1** [DMTF/libspdm#1406] [Ideas]
Ref: https://github.com/DMTF/libspdm/issues/1260

Ref: https://icmconference.org/wp-content/uploads/C22b-RuanX.pdf

# background 

## openssl

- Openssl 3.0 info
https://www.openssl.org/docs/fips.html
https://www.openssl.org/docs/man3.0/man7/fips_module.html

- Legacy Openssl 1.1 info
https://beta.openssl.org/docs/fips.html
https://beta.openssl.org/docs/fips/UserGuide-2.0.pdf

- NIST doc
https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4282.pdf

- API
https://github.com/openssl/openssl/blob/OpenSSL-fips-2_0-stable/fips/fips.h

```
unsigned long FIPS_module_version(void);
const char *FIPS_module_version_text(void);

int FIPS_module_mode_set(int onoff, const char *auth);
int FIPS_module_mode(void);

int FIPS_selftest(void);
int FIPS_selftest_failed(void);
```

## mbedtls

- info
https://mbed-tls.readthedocs.io/en/latest/kb/generic/is-mbedtls-fips-certified/
https://github.com/Mbed-TLS/mbedtls/issues/5995

## wolfssl

- info
https://www.wolfssl.com/license/fips/
https://www.wolfssl.com/docs/fips-ready-user-guide/

- NIST doc
https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp3389.pdf

- API
wolfssl-5.5.3-gplv3-fips-ready\wolfssl\wolfcrypt\fips.h

```
WOLFSSL_API const char* wolfCrypt_GetVersion_fips(void);
WOLFSSL_API int wolfCrypt_IntegrityTest_fips(void);

WOLFSSL_API int wolfCrypt_GetStatus_fips(void);
```

## TPM

- info
https://trustedcomputinggroup.org/resource/tcg-fips-140-2-guidance-for-tpm-2-0/

# proposal

## Design

- Cryptographic Algorithm Self-Test

Reference: 
   [1] Implementation Guide: https://csrc.nist.gov/CSRC/media/Projects/cryptographic-module-validation-program/documents/fips%20140-3/FIPS%20140-3%20IG.pdf
   [2] Cryptographic Algorithm Validation Program (CAVP): https://csrc.nist.gov/Projects/cryptographic-algorithm-validation-program
   [3] Cryptographic Module Validation Program (CMVP): https://csrc.nist.gov/Projects/cryptographic-module-validation-program
   [4] CMVP Approved Security Functions: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-140Cr1.pdf
   [5] CMVP Approved Sensitive Security Parameter Generation and Establishment Methods: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-140Dr1.pdf
   [6] Cryptographic Standards and Guidelines (Examples with Intermediate Values): https://csrc.nist.gov/projects/cryptographic-standards-and-guidelines/example-values

| SPDM Algorithm      | Specification                             | Type | Test Attributes      |
| ------------------- | ----------------------------------------- | ---- | -------------------- |
| SHA-256/384/512     | [NIST.FIPS.180-4](https://doi.org/10.6028/NIST.FIPS.180-4)   | KAT  | SHA-256/384/512          |
| SHA3-256/384/512    | [NIST.FIPS.202](https://doi.org/10.6028/NIST.FIPS.202)     | KAT  | SHA3-256/384/512            |
| RSA-SSA             | [NIST.FIPS.186-5](https://doi.org/10.6028/NIST.FIPS.186-5), [rfc8017](https://tools.ietf.org/html/rfc8017)   | KAT  | RSA-SSA2048 + SHA256 |
| RSA-PSS             | [NIST.FIPS.186-5](https://doi.org/10.6028/NIST.FIPS.186-5), [rfc8017](https://tools.ietf.org/html/rfc8017)   | KAT  | RSA-PSS2048 + SHA256 |
| ECDSA               | [NIST.FIPS.186-5](https://doi.org/10.6028/NIST.FIPS.186-5), [NIST.SP.800-186](https://doi.org/10.6028/NIST.SP.800-186)  | KAT  | ECDSA-P256+SHA256    |
| EdDSA               | [NIST.FIPS.186-5](https://doi.org/10.6028/NIST.FIPS.186-5), [NIST.SP.800-186](https://doi.org/10.6028/NIST.SP.800-186), [rfc8032](https://www.rfc-editor.org/rfc/rfc8032)  | KAT  | EdDSA-25519,EdDSA-448          |
| HMAC                | [NIST.FIPS.198-1](https://doi.org/10.6028/NIST.FIPS.198-1), [rfc2104](https://tools.ietf.org/html/rfc2104)   | KAT  | HMAC-SHA-256/384/512         |
| AES-GCM             | [NIST.FIPS.197](https://doi.org/10.6028/NIST.FIPS.197), [NIST.SP.800-38D](https://doi.org/10.6028/NIST.SP.800-38D)   | KAT  | AES-GCM-256          |
| FFDHE               | [NIST.SP.800-56Ar3](https://doi.org/10.6028/NIST.SP.800-56Ar3), [rfc7919](https://www.rfc-editor.org/rfc/rfc7919) | PCT  | FFDHE-2048           |
| ECDHE               | [NIST.SP.800-56Ar3](https://doi.org/10.6028/NIST.SP.800-56Ar3), [rfc8446](https://www.rfc-editor.org/rfc/rfc8446) | KAT  | ECDHE-P256           |
| HKDF                | [NIST.SP.800-56Cr2](https://doi.org/10.6028/NIST.SP.800-56Cr2), [rfc5869](https://tools.ietf.org/html/rfc5869) | KAT  | HKDF-HMAC-SHA-256    |
| ChaCha-Poly (*) | [rfc8439](https://www.rfc-editor.org/rfc/rfc8439) | KAT | not FIPS approved yet |
| SM3 (*) | [GB/T 32905-2016](http://www.gmbz.org.cn/upload/2018-07-24/1532401392982079739.pdf), [ISO/IEC 10118-3:2018](https://www.iso.org/standard/67116.html) | KAT | not FIPS approved yet |
| SM4-GCM (*) | [GB/T 32907-2016](http://www.gmbz.org.cn/upload/2018-04-04/1522788048733065051.pdf), [ISO/IEC 18033-3:2010/Amd 1:2021](https://www.iso.org/standard/81564.html), [rfc8998](https://tools.ietf.org/html/rfc8998) | KAT | not FIPS approved yet |
| SM2-digital-signature (\*) <br> SM2-key-exchange (\*) | [GB/T 32918.1-2016](http://www.gmbz.org.cn/upload/2018-07-24/1532401673134070738.pdf), [GB/T 32918.2-2016](http://www.gmbz.org.cn/upload/2018-07-24/1532401673138056311.pdf), [GB/T 32918.3-2016](http://www.gmbz.org.cn/upload/2018-07-24/1532401673149005052.pdf), <br> [GB/T 32918.4-2016](http://www.gmbz.org.cn/upload/2018-07-24/1532401673367034870.pdf), [GB/T 32918.5-2016](http://www.gmbz.org.cn/upload/2018-07-24/1532401863206085511.pdf), [ISO/IEC 14888-3:2018](https://www.iso.org/standard/76382.html) | KAT | not FIPS approved yet |
| SPDM-Key-Schedule (*) | [DMTF-DSP0274](https://www.dmtf.org/dsp/DSP0274)          | KAT  | not FIPS approved yet |

It maybe Known Answer Test (KAT) or Pairwise Consistency Test (PCT).

The Test Vector (KAT) can be found at [2] and [6].

- Software Integrity Self-Test

It uses HMAC-SHA-256, TBD.

## Implementation

- libspdm_lib_config macro

```
#ifdef LIBSPDM_FIPS_MODE
#define LIBSPDM_SM2_DSA_SUPPORT 0
#define LIBSPDM_EDDSA_ED25519_SUPPORT 0
#define LIBSPDM_EDDSA_ED448_SUPPORT 0
#define LIBSPDM_SM2_KEY_EXCHANGE_SUPPORT 0
#define LIBSPDM_AEAD_CHACHA20_POLY1305_SUPPORT 0
#define LIBSPDM_AEAD_SM4_SUPPORT 0
#define LIBSPDM_SM3_256_SUPPORT 0
#endif
```

- libspdm_common_lib API

```
#ifdef LIBSPDM_FIPS_MODE
const char* libspdm_get_module_version(void);
#endif
```

- libspdm_crypt_lib API

```
#ifdef LIBSPDM_FIPS_MODE
const char* libspdm_fips_get_module_version(void);
int libspdm_fips_run_selftest(void);
int libspdm_fips_get_selftest_result(void);
#endif
```

- libspdm_crypt_lib internal API

```
#ifdef LIBSPDM_FIPS_MODE
int libspdm_fips_module_integrity_test(void); // TBD

// sym
int libspdm_fips_selftest_sha2(void);
int libspdm_fips_selftest_sha3(void);
int libspdm_fips_selftest_hmac(void);
int libspdm_fips_selftest_hkdf(void);
int libspdm_fips_selftest_aes_gcm(void);

// asym digital signing (deterministic)
int libspdm_fips_selftest_rsa_ssa(void);
int libspdm_fips_selftest_eddsa(void);

// asym key exchange
int libspdm_fips_selftest_ffdhe(void);
int libspdm_fips_selftest_ecdhe(void);

// asym digital signing (probabilistic) - need hardcode random salt
int libspdm_fips_selftest_rsa_pss(void);
int libspdm_fips_selftest_ecdsa(void);

#endif
```

**Comment by xiaoyuruan** [comment]
- Need to add HMAC-SHA-512 KAT.
- Use some other mode for AES KAT, maybe AES-CBC-256.
- NIST PUB186-4 has RSASSA-PKCS1 v1.5 or RSASSA-PSS.  What does "RSA SSA" refer to?
- FIPS requires KAT on only one of RSASSA-PKCS1 v1.5 and RSASSA-PSS (both can still be performed; doesn't harm).
- FIPS does not require SPDM-Key-Schedule KAT (it can still be performed; doesn't harm).
- For ECDSA KAT and RSASSA-PSS KAT: hardcode random salt; check signature against hardcoded known answer too (not just sign then verify)

**Comment by jyao1** [comment]
"RSA SSA" in SPDM spec means "RSASSA-PKCS1 v1.5".

**Comment by jyao1** [comment]
Discussed with Xiaoyu, we are thinking to introduce another MACRO - `LIBSPDM_FIPS_TEST_AT_LOAD` for selftest.

1. If LIBSPDM_FIPS_TEST_AT_LOAD, then libspdm runs libspdm_fips_run_selftest() at load time.
2. If !LIBSPDM_FIPS_TEST_AT_LOAD, then libspdm runs individual libspdm_fips_selftest_xxx() when algo XXX is called and libspdm_fips_selftest_xxx() is not runs before.
3. The integrator can always call libspdm_fips_run_selftest() at any time, regardless LIBSPDM_FIPS_TEST_AT_LOAD value.

**Comment by jyao1** [comment]
Status summary:
1. FIPS SKU
  * Add FIPS 140-3 config for libspdm - https://github.com/DMTF/libspdm/pull/1576
2. zero key
  * Zero private key before free for FIPS - https://github.com/DMTF/libspdm/pull/1606
  * Zero key for FIPS when encode and decode - https://github.com/DMTF/libspdm/pull/1590
3. Only approved algorithms
  * Make only approved algo when in FIPS mode - https://github.com/DMTF/libspdm/pull/1577
  * Add check for FIPS approved algo - https://github.com/DMTF/libspdm/pull/1581
4. module_id API
  * Add fips_module_id api  - https://github.com/DMTF/libspdm/pull/1623
5. run_selftests API
  * Add aes_gcm selftest - https://github.com/DMTF/libspdm/pull/1647
  * Add rsa_ssa selftest - https://github.com/DMTF/libspdm/pull/1647
  * Add hmac self_test - https://github.com/DMTF/libspdm/pull/1647
  * Add hkdf self_test - https://github.com/DMTF/libspdm/pull/1678
  * Add ecdh self_test - https://github.com/DMTF/libspdm/pull/1679
  * Add sha2 self_test - https://github.com/DMTF/libspdm/pull/1680
  * Add sha3 self_test - https://github.com/DMTF/libspdm/pull/1687
  * Add ffdh self_test - https://github.com/DMTF/libspdm/pull/1695
  * Add ecdsa self_test - https://github.com/DMTF/libspdm/pull/1681
  * Add eddsa self_test - https://github.com/DMTF/libspdm/pull/1803
  * Add rsa_pss self_test - https://github.com/DMTF/libspdm/pull/2277
6. run selftest_results
  * Add FIPS selftest and get_result_selftest - https://github.com/DMTF/libspdm/pull/1647
7. integrity selftest.
  * Defer
8. pre-operational selftest.
  * libspdm_fips_run_selftest - https://github.com/DMTF/libspdm/pull/1833
  * FIPS selftest refactor -https://github.com/DMTF/libspdm/pull/1974
